x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java

```diff
@@ -341,6 +343,35 @@ protected void masterOperation(
         );
     }
 
+    void verifyMlNodesAndModelArchitectures(TrainedModelConfig configToReturn, Client client, ThreadPool threadPool) {
+        ActionListener<Void> failureListener = new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void o) {}
+
+            @Override
+            public void onFailure(Exception e) {
+                HeaderWarning.addWarning(e.getMessage());
```

Comment:
And change this to pass `configToReturn` to a listener passed as an argument _after_ adding the header warning.